drone/src/agent/backend.rs

```diff
@@ -0,0 +1,145 @@
+use std::{net::IpAddr, time::Duration};
+
+use super::docker::DockerInterface;
+use anyhow::{anyhow, Result};
+use dis_spawner::{
+    messages::{
+        agent::{BackendStatsMessage, DroneLogMessage},
+        dns::{DnsRecordType, SetDnsRecord},
+    },
+    nats::TypedNats,
+    types::{BackendId, ClusterName},
+};
+use tokio::{task::JoinHandle, time::sleep};
+use tokio_stream::StreamExt;
+
+/// JoinHandle does not abort when it is dropped; this wrapper does.
+struct AbortOnDrop<T>(JoinHandle<T>);
+
+impl<T> Drop for AbortOnDrop<T> {
+    fn drop(&mut self) {
+        self.0.abort();
+    }
+}
+
+pub struct BackendMonitor {
+    _log_loop: AbortOnDrop<Result<(), anyhow::Error>>,
+    _stats_loop: AbortOnDrop<Result<(), anyhow::Error>>,
+    _dns_loop: AbortOnDrop<Result<(), anyhow::Error>>,
+}
+
+impl BackendMonitor {
+    pub fn new(
+        backend_id: &BackendId,
+        cluster: &ClusterName,
+        ip: IpAddr,
+        docker: &DockerInterface,
+        nc: &TypedNats,
+    ) -> Self {
+        let log_loop = Self::log_loop(backend_id, docker, nc);
+        let stats_loop = Self::stats_loop(backend_id, docker, nc);
+        let dns_loop = Self::dns_loop(backend_id, ip, nc, cluster);
+
+        BackendMonitor {
+            _log_loop: AbortOnDrop(log_loop),
+            _stats_loop: AbortOnDrop(stats_loop),
+            _dns_loop: AbortOnDrop(dns_loop),
```

Comment:
So, I understand this, but how do you know that these event loops get moved instead of copied? Here it's obvious, but I'm not entirely clear on the mechanism for this. Is it that unless .clone() is called, a non borrow is always a move?

Reply:
Yes, if a value does not implement `Copy`, it has to be cloned explicitly by calling `.clone()`. `JoinHandle`, the underlying type which holds the threads, not only doesn't implement `Copy`, it also does not implement `Clone`. This means that _can't_ be cloned, even explicitly. So if we own a `JoinHandle`, we own the only handle to that task.

(If we wanted to access it in multiple places, we could wrap it in an `Arc` or `Rc`.)